validator/find.go

```diff
@@ -0,0 +1,109 @@
+/*
+# Copyright 2024 NVIDIA CORPORATION
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+*/
+
+package main
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+)
+
+type root string
+
+// getDriverLibraryPath returns path to `libnvidia-ml.so.1` in the driver root.
+// The folder for this file is also expected to be the location of other driver files.
+func (r root) getDriverLibraryPath() (string, error) {
+	librarySearchPaths := []string{
+		"/usr/lib64",
+		"/usr/lib/x86_64-linux-gnu",
+		"/usr/lib/aarch64-linux-gnu",
+		"/lib64",
+		"/lib/x86_64-linux-gnu",
+		"/lib/aarch64-linux-gnu",
+	}
+
+	libraryPath, err := r.findFile("libnvidia-ml.so.1", librarySearchPaths...)
+	if err != nil {
+		return "", err
+	}
+
+	return libraryPath, nil
+}
+
+// getNvidiaSMIPath returns path to the `nvidia-smi` executable in the driver root.
+func (r root) getNvidiaSMIPath() (string, error) {
+	binarySearchPaths := []string{
+		"/usr/bin",
+		"/usr/sbin",
+		"/bin",
+		"/sbin",
+	}
+
+	binaryPath, err := r.findFile("nvidia-smi", binarySearchPaths...)
+	if err != nil {
+		return "", err
+	}
+
+	return binaryPath, nil
+}
+
+// isDevRoot checks whether the specified root is a dev root.
+// A dev root is defined as a root containing a /dev folder.
+func (r root) isDevRoot() bool {
+	stat, err := os.Stat(filepath.Join(string(r), "dev"))
+	if err != nil {
+		return false
+	}
+	return stat.IsDir()
+}
+
+// getDevRoot returns the dev root associated with the root.
+// If the root is not a dev root, this defaults to "/".
+func (r root) getDevRoot() string {
+	if r.isDevRoot() {
+		return string(r)
+	}
+	return "/"
+}
```

Comment:
I was surprised to not see a field for `devPath` in `HostPathsSpec`, but then I recalled a conversation about not needing to specify it if we assume that it is either only ever at:
1. `<DriverInstallDir>/dev`; OR
1. `/dev`

It looks like this is the logic that is implementing this assumption.

Reply:
Correct. We decided to omit `devPath` from `HostPathsSpec` to start with. If a use case presents itself where device nodes need to exist at some other location, we can introduce this field to the API.